learn/shorts/functional-programming-theory.qmd

```diff
@@ -0,0 +1,202 @@
+
+In the context of functional programming, keeping data structures
+immutable is preferred because they help avoid side effects. The code
+written should strive to avoid changing the data it is working with.
+Instead, it should create new data structures based on the existing
+ones. This makes the code easier to reason about and less prone to bugs.
+
+For example, using immutable data structures ensures that functions do
+not introduce subtle bugs due to unintended modifications. Consider the
+difference between modifying a list in place versus creating a new list
+with modified values:
+
+``` python
+# Mutable list modification
+my_list = [1, 2, 3]
+my_list.append(4)  # This modifies the original list
+
+# Immutable approach using tuples
+my_tuple = (1, 2, 3)
+new_tuple = my_tuple + (4,)  # This creates a new tuple instead of modifying the original
+```
+
+### What is a side effect in programming?
+
+A side effect is any change in the state of the system that is
+observable outside the called function other than the return value. This
+can include modifying a variable outside the function, printing to the
+console, or modifying a data structure.
+
+Side effects can make code harder to understand, as they introduce
+hidden dependencies and make it difficult to predict the behavior of the
+code. Functional programming aims to minimize side effects by using
+immutable data structures and pure functions.
+
+For example, the following function has a side effect because it
+modifies an external variable:
+
+``` python
+counter = 0
+
+def increment():
+    global counter  # Modifies an external variable (side effect)
+    counter += 1
+```
+
+A functional programming approach would avoid modifying external state:
+
+``` python
+def increment(counter):
+    return counter + 1  # No side effects, purely functional
+```
+
+### What is a pure function?
+
+A pure function is a fundamental concept in functional programming, it
+is basically a function that has no side effects and always returns the
+same output for the same input. This means that a the ideal pure
+function does not modify any external state and does not depend on any
+external state.
+
+Pure functions are easier to test, reason about, and parallelize, as
+they do not have hidden dependencies or side effects. They are also
+easier to reuse, as they can be called with the same arguments and will
+always return the same result.
+
+For example, the following function is pure:
+
+``` python
+def add(a, b):
+    return a + b  # No side effects, always returns the same output for the same input
+```
+
+By contrast, the following function is impure because it modifies an
+external variable:
+
+``` python
+total = 0
+
+def add_to_total(n):
+    global total
+    total += n  # Modifies external state (side effect)
+```
+
+### What is a higher-order function?
+
+A higher-order function is a function which will take one or more
+existing functions as arguments, and/or a function which returns a new
```

Comment:
Highlight that higher-order functions (also known as functionals) are predictable in their output, so that functions like map and filter always output the same number of objects as is given it, while reduce always outputs one object, no matter how many items you give it.